tests/unit/lms/views/api/canvas/sync_test.py

```diff
@@ -1,140 +1,148 @@
+    user_id = request_json["learner"]["canvas_user_id"]
+    canvas_api_client.course_sections.assert_called_once_with(course_id)
+    canvas_api_client.users_sections.assert_called_once_with(user_id, course_id)
+
+    assert_sync_and_return(
+        groupids, synced_sections=sections.course, returned_sections=sections.user,
     )
-    assert returned_groupids == [
-        "group:d99674b9700f4a40a2b301d2949b61339c58236c@TEST_AUTHORITY"
-    ]
 
 
-def assert_that_it_called_sync_and_returned_the_groupids(
-    lti_h_service, returned_groupids, expected_groups
-):
-    lti_h_service.sync.assert_called_once_with(expected_groups, {"foo": "bar"})
-    assert returned_groupids == [
-        f"group:{group.authority_provided_id}@TEST_AUTHORITY"
-        for group in expected_groups
-    ]
+pytestmark = pytest.mark.usefixtures("canvas_api_client", "lti_h_service")
 
 
-pytestmark = pytest.mark.usefixtures("canvas_api_client", "lti_h_service")
+@pytest.fixture
+def assert_sync_and_return(lti_h_service, request_json):
+    tool_guid = request_json["lms"]["tool_consumer_instance_guid"]
+    context_id = request_json["course"]["context_id"]
+
+    def _groups_for(sections):
+        return [
+            HGroup.from_lti_parts(
+                name=section.get("name", f"Section {section['id']}"),
+                tool_consumer_instance_guid=tool_guid,
+                context_id=context_id,
+                section_id=section["id"],
+                type_="section_group",
+            )
+            for section in sections
+        ]
+
+    def assert_return_values(groupids, synced_sections, returned_sections):
+        lti_h_service.sync.assert_called_once_with(
+            _groups_for(synced_sections), request_json["group_info"]
+        )
+
+        assert groupids == [
+            group.groupid(TEST_SETTINGS["h_authority"])
+            for group in _groups_for(returned_sections)
+        ]
+
+    return assert_return_values
 
 
 @pytest.fixture
-def canvas_api_client(canvas_api_client):
-    # The course has three sections named "Section 1", "Section 2" and "Section
-    # 3" (with IDs 1, 2 and 3).
-    canvas_api_client.course_sections.return_value = [
-        {"id": i, "name": f"Section {i}",} for i in range(1, 4)
-    ]
+def is_instructor(pyramid_request):
+    pyramid_request.lti_user = pyramid_request.lti_user._replace(roles="Instructor")
 
-    # The learner is only a member of section 2.
-    canvas_api_client.authenticated_users_sections.return_value = [
-        {"id": 2, "name": "Section 2"}
-    ]
 
-    # users_sections() returns id's only, not names.
-    canvas_api_client.users_sections.return_value = [{"id": 2}]
+@pytest.fixture
+def is_learner(pyramid_request):
+    pyramid_request.lti_user = pyramid_request.lti_user._replace(roles="Learner")
+
+
+@pytest.fixture
+def is_speedgrader(request_json):
+    request_json["learner"] = {"canvas_user_id": "user_id"}
+
+
+@pytest.fixture
+def sections():
+    class Sections:
```

Comment:
Put all the section data into a fixture that we can directly refer to in the tests. This is another part of making this easier to understand the flow of.

Previously we had these values in the `canvas_api_client` alone, and then we constructed separate groups that happened to correspond to them with hash values. 

Now we pass these around directly so instead of creating a set of groups we say `_groups_for(sections.user)` etc.